### ISSUE TITLE
[Parity 7/16] Data loading - Dataset/DataLoader/Samplers/collate/workers/prefetch/datapipes

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#216]
Part of epic #209.

## Summary

PyTorch's `torch.utils.data` is the de-facto standard for ML data pipelines. Our library has `PredictionModelBuilder` at the facade but no tensor-native data-loading primitives. This blocks large-scale training workflows, streaming datasets, and deterministic data pipelines.

### Core
- `Dataset` (map-style: `__getitem__`, `__len__`)
- `IterableDataset` (stream / generator style)
- `ConcatDataset`, `ChainDataset`, `Subset`, `random_split`

### Loader
- `DataLoader(dataset, batch_size, shuffle, num_workers, pin_memory, drop_last, collate_fn, worker_init_fn, persistent_workers, prefetch_factor, timeout, multiprocessing_context, sampler, batch_sampler, generator)`
- Multi-process worker pool with shared-memory tensor transfer
- Automatic collation (stack tensors, pad variable shapes)

### Samplers
- `RandomSampler`, `SequentialSampler`, `WeightedRandomSampler`, `SubsetRandomSampler`, `BatchSampler`, `DistributedSampler`

### DataPipes (`torchdata`)
- Functional composable ops (`.map`, `.filter`, `.batch`, `.shuffle`, `.bucket_batch`, `.sharding_filter`, `.zip`, `.concat`, `.demultiplex`, `.mux`)
- Graph-based pipeline optimization
- Sharding filter for distributed workers

## Current state in the repo

- Nothing in the Tensors repo. Data loading likely lives in upstream AiDotNet core.
- `PredictionModelBuilder` accepts training data as arrays; no streaming, no workers, no prefetch.

## How we beat PyTorch

1. **Zero-copy worker transfer**: on Windows where `fork()` isn't available, PyTorch's `num_workers>0` serializes through pickle. We use .NET `Memory<T>` + shared-memory files + pointer passing via `System.IO.MemoryMappedFiles`. Target 5–10× on Windows multi-worker throughput.
2. **GPU-direct prefetch**: pipeline stages that decode + augment directly on-GPU via the existing `GpuStreamPool` — PyTorch always decodes on CPU then copies. Our model: decode on CPU, augment on GPU, `pin_memory=True` auto-managed.
3. **Deterministic by default**: seed propagation per (epoch, worker, sample-index) via Philox counters (from the distributions issue). PyTorch requires `worker_init_fn` boilerplate.
4. **First-class `CheckpointableLoader`** that can resume mid-epoch after a crash — PyTorch's `StatefulDataLoader` is in `torchdata` beta.
5. **Memory-format aware collation**: batched output can be NCHWc or channels-last without a re-layout. PyTorch always outputs NCHW.
6. **Compiled pipeline**: DataPipe graphs can be captured by `LazyTensorScope` so decode + augment fuse with the first model op.
7. **Sub-byte batch packing**: when loading quantized embeddings or int4 weights, the loader emits packed tensors directly. PyTorch loads fp32 then quantizes.
8. **Cross-platform worker pool**: `ThreadPool` + `Task` primitives that work identically on Windows / Linux / macOS. PyTorch workers are Linux-optimized.

### Datasets
- [ ] `IDataset<T>` (map-style: `this[int idx]`, `Count`)
- [ ] `IIterableDataset<T>` (stream)
- [ ] `ConcatDataset`, `ChainDataset`, `Subset`, `RandomSplit`
- [ ] `TensorDataset<T>` (from in-memory tensors)
- [ ] `InMemoryDataset<TSample>`, `OnDiskDataset<TSample>` presets
- [ ] `CachedDataset<T>` with LRU caching of decoded samples

### Samplers
- [ ] `RandomSampler`, `SequentialSampler`, `WeightedRandomSampler`
- [ ] `SubsetRandomSampler`, `BatchSampler`
- [ ] `DistributedSampler` (integrates with the distributed issue)

### DataLoader
- [ ] `DataLoader<T, TBatch>` with async prefetch, worker pool, pin_memory, drop_last
- [ ] Multi-worker shared-memory transfer (Windows + Linux + macOS)
- [ ] Configurable `collate_fn` with defaults for tensor stacking + padding
- [ ] `persistent_workers`, `prefetch_factor`, `timeout`
- [ ] Seed propagation per (epoch, worker, sample) via Philox

### DataPipes
- [ ] `IDataPipe<T>.Map<U>`, `.Filter`, `.Batch`, `.Shuffle`, `.BucketBatch`, `.Zip`, `.Concat`, `.Mux`, `.Demultiplex`
- [ ] `.ShardingFilter` for distributed workers
- [ ] Graph-based optimization pass (fuses Map chains, reorders Filter before Map when predicate doesn't need the mapped value)

### Bonus
- [ ] `CheckpointableLoader<T>` — resume mid-epoch
- [ ] `GpuPrefetch` — decode on CPU, augment on GPU
- [ ] `PackedSampleLoader<T>` — sub-byte / FP8 tensor loading without full dequant

## Acceptance criteria

- Multi-worker correctness test: same dataset, different worker counts, output set is identical.
- Determinism test: same seed → same iteration order, regardless of worker count.
- Windows multi-worker benchmark: ≥5× PyTorch `num_workers=4` on a standard image-decode pipeline.
- GpuPrefetch benchmark: overlaps H2D with compute; no CPU stall on a ResNet-50 training step.
- `DistributedSampler` integration test with the process group from the distributed issue.
- Checkpoint mid-epoch + resume produces identical subsequent iteration order.

## Implementation notes

- New module `src/AiDotNet.Tensors/Data/`.
- Workers use `System.Threading.Channels` + `System.IO.MemoryMappedFiles` for zero-copy.
- Pin_memory via `NativeMemoryManager.cs`.
- DataPipe graph reuses `LazyTensorScope` / `ILazyGraphOptimizationPass` patterns.
- `TensorDataset<T>` and `TensorDataLoader<T>` surface on `PredictionModelBuilder` for the end-user path.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

Nothing in this repo. `PredictionModelBuilder` accepts array-style training data but no workers/prefetch.

## 2. Scope — full inventory

### Datasets
```
IDataset<T> (map-style: this[int idx], Count)
IIterableDataset<T> (stream)
ConcatDataset, ChainDataset, Subset, RandomSplit
TensorDataset<T> (from in-memory tensors)
InMemoryDataset<TSample>, OnDiskDataset<TSample> presets
CachedDataset<T> with LRU caching
```

### Samplers
```
RandomSampler, SequentialSampler, WeightedRandomSampler
SubsetRandomSampler, BatchSampler
DistributedSampler (integrates with #215)
```

### DataLoader
```
DataLoader<T, TBatch>
  - async prefetch, worker pool, pin_memory, drop_last
  - multi-worker shared-memory transfer (Win/Linux/macOS)
  - configurable collate_fn with defaults (tensor stacking + padding)
  - persistent_workers, prefetch_factor, timeout
  - seed propagation per (epoch, worker, sample) via Philox (#213)
```

### DataPipes
```
IDataPipe<T>.{Map<U>, Filter, Batch, Shuffle, BucketBatch, Zip, Concat, Mux, Demultiplex,
              ShardingFilter}
Graph-based optimization pass (fuses Map chains, reorders Filter-before-Map when safe)
```

### Bonus
```
CheckpointableLoader<T> — resume mid-epoch
GpuPrefetch — decode on CPU, augment on GPU
PackedSampleLoader<T> — sub-byte / FP8 loading without full dequant
```

**Total: ~30 surfaces.**

## 3. Win vectors

1. **Zero-copy Windows worker transfer**: `Memory<T>` + `System.IO.MemoryMappedFiles` — target 5–10× PyTorch `num_workers=4` on Windows (pickle serialization elimination).
2. **GPU-direct prefetch**: decode on CPU, augment on GPU via `GpuStreamPool`.
3. **Deterministic by default**: Philox per-(epoch, worker, sample) seeding — no `worker_init_fn` boilerplate.
4. **First-class `CheckpointableLoader`** — PyTorch's `StatefulDataLoader` is torchdata beta.
5. **Memory-format aware collation** — output NCHWc/channels-last directly.
6. **Compiled pipeline**: DataPipe graphs captured by `LazyTensorScope` — decode + augment fuse with first model op.
7. **Sub-byte batch packing**: emit packed tensors directly (int1/2/3/4/NF4/FP4/FP8).
8. **Cross-platform worker pool** — same behavior on Win/Linux/macOS.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Windows multi-worker image-decode throughput | ≥5× PyTorch `num_workers=4` |
| Linux multi-worker throughput | ≥ parity PyTorch |
| Determinism | same seed + different worker count → identical order |
| GPU prefetch overlap | H2D overlaps compute on ResNet-50 step (no CPU stall) |
| Checkpoint-resume mid-epoch | bit-identical subsequent iteration order |
| Sub-byte batch loading | works without dequant step |
| DistributedSampler (#215) | balanced partition across ranks |

## 5. File inventory

**New (~60):**
- `src/AiDotNet.Tensors/Data/` root
- `Data/Datasets/` — IDataset, IIterableDataset, TensorDataset, Concat, Chain, Subset, RandomSplit, InMemory, OnDisk, Cached (~15)
- `Data/Samplers/` — Random, Sequential, WeightedRandom, SubsetRandom, Batch, Distributed (~8)
- `Data/Loader/` — DataLoader, WorkerPool, SharedMemoryTransfer, PinMemory, Collate, CheckpointableLoader (~15)
- `Data/Pipes/` — IDataPipe + Map/Filter/Batch/Shuffle/BucketBatch/Zip/Concat/Mux/Demux/ShardingFilter + DataPipeGraphOptimizer (~15)
- `Data/Helpers/` — PhiloxSeeding, GpuPrefetch, PackedSampleLoader (~5)

**Modified (~5):** `NativeMemoryManager.cs` (pin_memory), `LazyTensorScope.cs` (DataPipe capture), `CompiledInferencePlan.cs` (prefetch hook), `PredictionModelBuilder.cs` (surface).

**Tests (~25):** multi-worker correctness, determinism, Windows perf, GpuPrefetch overlap, DistributedSampler integration, checkpoint-resume bit-identical.

**Benchmarks (~5):** `WindowsMultiWorkerBenchmarks`, `LinuxMultiWorkerBenchmarks`, `GpuPrefetchBenchmarks`, `DataPipeOptimizationBenchmarks`, `SubByteLoadingBenchmarks`.

## 6. Acceptance gate

1. Multi-worker correctness test (same set across worker counts).
2. Windows ≥5× PyTorch `num_workers=4`.
3. GPU prefetch no-CPU-stall demonstrated on ResNet-50.
4. Checkpoint mid-epoch → resume bit-identical.
5. DistributedSampler integration works with #215.
6. Sub-byte loader works without dequant.
7. DataPipe graph optimization pass demonstrable speedup (fuses Map chain).

## 7. Timeline — ~4 weeks

W1 IDataset / IIterableDataset / presets + samplers.
W2 DataLoader + worker pool + SharedMemoryTransfer + pin_memory + collate.
W3 DataPipes + graph optimization + CheckpointableLoader + GpuPrefetch.
W4 Benchmarks + PR.

## 8. Dependencies

- **Blocks**: end-to-end training demos.
- **Blocked by**: #213 (Philox for seeding), #215 (DistributedSampler needs IProcessGroup).

## 9. Risks

- **Windows process boundary** — `Memory<T>` across processes needs MMF; design + test carefully.
- **macOS fork vs spawn** — ship `ThreadPool` + `Task` that sidesteps the issue.
- **DataPipe graph optimization** — pattern-matching correctness; extensive unit tests.

---

*Part of epic #209. Plan ratified 2026-04-19.*
